src/samples/mqtt/bluemix.c

```diff
@@ -0,0 +1,355 @@
+}
+
+static void
+on_disconnect(void *data, struct sol_mqtt *mqtt)
+{
+    SOL_INF("Reconnecting...");
+    sol_timeout_add(1000, try_reconnect, mqtt);
+}
+
+static struct sol_mqtt_config config = {
+    .api_version = SOL_MQTT_CONFIG_API_VERSION,
+    .clean_session = true,
+    .keepalive = 60,
+    .username = "use-token-auth",
+    .handlers = {
+        .connect = on_connect,
+        .disconnect = on_disconnect,
+    },
+};
+
+static bool
+mqtt_init(void *vdata)
+{
+    struct bluemix_data *data = (struct bluemix_data *)vdata;
+    struct sol_buffer buffer = SOL_BUFFER_INIT_EMPTY;
+    int r;
+
+    SOL_NULL_CHECK_GOTO(data, error);
+
+    r = sol_buffer_append_printf(&buffer, "%s.messaging.internetofthings.ibmcloud.com", data->org_id);
+    SOL_INT_CHECK_GOTO(r, < 0, error);
+
+    config.client_id = data->client_id;
+    config.password = data->auth_token;
+
+    data->mqtt = sol_mqtt_connect(buffer.data, 1883, &config, data);
+
+    sol_buffer_fini(&buffer);
+
+    if (!data->mqtt) {
+        SOL_WRN("Unable to create MQTT session. Retrying...");
+        return true;
+    }
+
+    SOL_INF("Sending sensor data");
+
+    return false;
+
+error:
+    SOL_WRN("Unable to initialize MQTT");
+    sol_quit();
+
+    return false;
+}
+
+/**
+ * Authenticate to HTTP Server
+ */
+
+static bool
+json_token_to_string(struct sol_json_token *token, char **out)
+{
+    if (sol_json_token_get_type(token) != SOL_JSON_TYPE_STRING)
+        return false;
+    free(*out);
+    *out = strndup(token->start + 1, token->end - token->start - 2);
+    return !!*out;
+}
+
+static bool
+parse_auth_json(struct bluemix_data *data, const char *json, size_t size)
+{
+    struct sol_json_scanner scanner;
+    struct sol_json_token token, key, value;
+    enum sol_json_loop_reason reason;
+
+    sol_json_scanner_init(&scanner, json, size);
+
+    SOL_JSON_SCANNER_OBJECT_LOOP (&scanner, &token, &key, &value, reason) {
+        if (SOL_JSON_TOKEN_STR_LITERAL_EQ(&key, "clientid")) {
+            if (!json_token_to_string(&value, &data->client_id))
+                goto error;
+        } else if (SOL_JSON_TOKEN_STR_LITERAL_EQ(&key, "authtoken")) {
+            if (!json_token_to_string(&value, &data->auth_token))
+                goto error;
+        } else if (SOL_JSON_TOKEN_STR_LITERAL_EQ(&key, "orgid")) {
+            if (!json_token_to_string(&value, &data->org_id))
+                goto error;
+        }
+    }
+
+    return true;
+
+error:
+    SOL_WRN("Error parsing auth json");
+    return false;
+}
+
+static bool get_user_token(void *vdata);
+
+static void
+request_callback(void *vdata, const struct sol_http_client_connection *connection,
+    struct sol_http_response *response)
+{
+    struct bluemix_data *data = (struct bluemix_data *)vdata;
+
+    if (!response->content.used || response->response_code != SOL_HTTP_STATUS_OK) {
+        SOL_DBG("Message: %.*s", (int)response->content.used, (char *)response->content.data);
```

Comment:
I don't think really think SOL_DBG("Message: %.*s", SOL_STR_SLICE_PRINT(sol_buffer_get_slice(response->content))); is that much simpler... but anyway, I will update.